src/zenml/models/v2/core/project.py

```diff
@@ -0,0 +1,203 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Models representing projects."""
+
+import re
+from typing import Any, Dict, Optional
+
+from pydantic import Field, model_validator
+
+from zenml.constants import STR_FIELD_MAX_LENGTH, STR_ID_FIELD_MAX_LENGTH
+from zenml.models.v2.base.base import (
+    BaseDatedResponseBody,
+    BaseIdentifiedResponse,
+    BaseRequest,
+    BaseResponseMetadata,
+    BaseResponseResources,
+    BaseUpdate,
+)
+from zenml.models.v2.base.filter import BaseFilter
+from zenml.utils.pydantic_utils import before_validator_handler
+
+# ------------------ Request Model ------------------
+
+
+class ProjectRequest(BaseRequest):
+    """Request model for projects."""
+
+    name: str = Field(
+        title="The unique name of the project. The project name must only "
+        "contain only lowercase letters, numbers, underscores, and hyphens and "
+        "be at most 50 characters long.",
+        min_length=1,
+        max_length=STR_ID_FIELD_MAX_LENGTH,
+        pattern=r"^[a-z0-9_-]+$",
+    )
+    display_name: str = Field(
+        default="",
+        title="The display name of the project.",
+        max_length=STR_FIELD_MAX_LENGTH,
+    )
+    description: str = Field(
+        default="",
+        title="The description of the project.",
+        max_length=STR_FIELD_MAX_LENGTH,
+    )
+
+    @model_validator(mode="before")
+    @classmethod
+    @before_validator_handler
+    def _validate_project_name(cls, data: Dict[str, Any]) -> Dict[str, Any]:
+        """Validate the project name.
+
+        Args:
+            data: The values to validate.
+
+        Returns:
+            The validated values.
+        """
+        name = data.get("name")
+        display_name = data.get("display_name")
+
+        if not name and not display_name:
+            return data
```

Comment:
This will trigger a validation error which i guess is intended?

Reply:
Yes :smile: , I just pass it to Pydantic here.